src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/CorrelationVector.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+
+namespace System.Diagnostics
+{
+    /// <summary>
+    /// This class represents a lightweight vector for identifying and measuring
+    /// causality.
+    /// </summary>
+    public class CorrelationVector
+    {
+        private const byte BaseLength = 22;
+        private const byte MaxVectorLength = 127;
+
+        private const char ElementChar = '.';
+        private const char ResetChar = '#';
+        private const char SpinChar = '_';
+
+        private string baseVector = null;
+        private int extension = 0;
+        private object resetLock = new object();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CorrelationVector"/> class.
+        /// </summary>
+        public CorrelationVector()
+        {
+            this.baseVector = CorrelationVector.GetBase(isReset: false);
+        }
+
+        /// <summary>
+        /// Gets the previous value of the Vector Base prior to a reset having been performed.
+        /// Value will be null if reset has not yet been performed.
+        /// </summary>
+        public string PreviousBase { get; private set; } = null;
+
+        /// <summary>
+        /// Gets the value of the Correlation Vector as a string.
+        /// </summary>
+        public string Value
+        {
+            get
+            {
+                return string.Concat(this.baseVector, ElementChar, this.extension);
+            }
+        }
+
+        /// <summary>
+        /// Increments the current extension by one. Do this before passing the value to an
+        /// outbound message header.
+        /// </summary>
+        /// <returns>
+        /// The new value as a string that you can add to the outbound message header.
+        /// </returns>
+        public string Increment()
+        {
+            int snapshot = 0;
+            int next = 0;
+            do
+            {
+                snapshot = this.extension;
+                if (snapshot == int.MaxValue)
+                {
+                    // Don't proceed past int.MaxValue, since we don't have a performant way of reliably
+                    // incrementing and rolling-over (Interlocked.CompareExchange doesn't have a uint implementation)
+                    return this.Value;
+                }
+                next = snapshot + 1;
+                int size = baseVector.Length + 1 + (int)Math.Log10(next) + 1;
+                if (size > CorrelationVector.MaxVectorLength)
+                {
+                    // Perform a reset
+                    lock (resetLock)
```

Comment:
I didn't see another way to make sure only one thread resets the Base without using a lock. The lock will only be acquired if the reset needs to happen, as opposed to on every increment. Do you have concerns with this? Or suggestions for an alternate approach?